TritonSE/TAP-Foster-Dog-Matching
# Application progress ping pong between admin and fosters

### Administrative Info
[GitHub Issue #68 ]

### Changes
- set up the back and forth up between admin and user
- when the user modifies their application/completes their actions items (like scheduling an interview), that is reflected on the admin side; when an admin gives approval/rejection or something else, it is reflected on the application state of the user

### Testing Instructions
- Create a new foster user in one tab and in an incognito tag (firebase uses cookies) log into the management account (management@tap.test)
- Create a new application on the foster side and fill it out
- On the management side, go to the application you just created; it should be at the very bottom of the pending applications list
- Approve/reject it 
- Complete next stage on the foster side and approve/reject it on the management side (repeat this step until you are at the very end)
- When doing the above step make sure that the behavior is as you'd expect for each substep
- If you need testing guidance message me on slack because this is a big pr to test

### Notes
- If at any point you want to go back in an application, you can manually go to mongodb, find your application document and modify the `status` field. The possible values are: `"Step 1: Initial Screen"`, `"Step 2: Initial Interview"`, `"Step 3: Home Screen"`, `"Step 4: Foster Matching"`, `"Step 5: Meet & Greet"`, `"Step 6: Foster in Home"`, `"rejected"`.
- The `messages` field on the application object contains the messages that the admin gives to the user. On stage 4, when `messages.stage4` is populated by the admin, it'll make the user go to a certain stage on the substep, so if you want to backtrack on that substep, you'll have to delete `messages.stage4` from your mongodb document.
- As mentioned on slack, there are issues on the admin side of the foster matching step. We don't store timestamps or have an efficient way of checking other people's preferences on a dog. 
- When a dog is fostered, because of the issues on step 4 admin side, we are not updating that dogs status on the backend to "in home"; the dog is left as is, so the inventory on the manage dog profiles page will not reflect successful adoptions.
- Step 6 on the admin side looks strange


### Update
- All issues mentioned in the notes above should be resolved, besides the last one, which I believe is in Andrew's pr.
- ALL content is now dynamically pulled, and once a foster makes it through the whole application process successfully, the dog that they end up matching with ends up "in home" (can see that update is correctly reflected in manage dog profiles page on the management role).
- In order to prevent the bug where the interview route 400s when an ambassador is not assigned to an app, I made it so that the admin can only click pass on stage 1 if the app they are reviewing is assigned to an ambassador. If it isn't a red error message prompting them to assign an admin shows up.
- The timestamps and other people's preferences mentioned in the notes above is fully omitted now, as per discussion
